[tf.data service] Test split provider failures with distributed and zipped datasets

kvignesh1420 opened the pull request:
This PR extends the test cases for `data_service_ops` by adding the following tests:
- Test the failing case of creating a split provider for a `ZipDataset`
- Test the failing case of creating a split provider for a `DataServiceDataset`.

cc: @aaudiber

kvignesh1420 committed:
test processing mode combinations with two level distributions

kvignesh1420 committed:
lint fixes

kvignesh1420 committed:
preserve error regex over pylint

kvignesh1420 committed:
distribute already distributed datasets

kvignesh1420 committed:
use dataset_ops instead of tf import

kvignesh1420 committed:
parameterize processing_mode

kvignesh1420 committed:
choose one mode at a time

kvignesh1420 committed:
split the combinations into different tests

kvignesh1420 committed:
remove empty lines

kvignesh1420 committed:
change function name as per review

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
+      self.getDatasetOutput(ds)
+
+  @combinations.generate(test_base.eager_only_combinations())
+  def testDistributeDistributedToParallel(self):
```
@aaudiber I have observed that the `testTwoLevelDistribute` test covers distributing a `parallel_epochs` based dataset using another `parallel_epochs` dataset. However, out of the 4 four test combinations that are possible with `parallel_epochs` and `distributed_epoch` when performing a two-level distribute were missing. Thus, I thought of adding this test and the one above.

aaudiber replied:
I think the tests would be useful if there was an interesting interaction between processing modes, but there isn't. The only interaction is that `distributed_epoch` requires that its input is splittable, and distributed datasets are not splittable (regardless of whether they use `parallel_epochs` or `distributed_epochs`). `parallel_epochs` doesn't care whether its input dataset is distributed or not, much less what processing mode its input dataset is using.

kvignesh1420 replied:
I agree. I have removed the redundant tests to address this.

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
+    self.assertDatasetProduces(
+        ds, [i + 1 for i in numbers], assert_items_equal=True)
+
+  def testParallelZippedDistributedDatasets(self):
```
@aaudiber apologies if the name of the test is confusing. However, since I didn't find any test that tries to distribute a zipped dataset which was prepared using already distributed datasets, I thought of adding this. Will remove it if you feel that this is redundant.

aaudiber replied:
No worries. I think we need to consider the cost of maintaining and running these tests during the build, so it's not worth adding this specific test when there are already such similar tests

kvignesh1420 replied:
I agree. Removed the redundant test.

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
+    ds3 = self.make_distributed_dataset(
+        ds3, cluster_3, processing_mode="distributed_epoch")
+
+    error_regex = "Cannot create a split provider for dataset of type ZipDataset"  # pylint: disable=line-too-long
```
@aaudiber  I was only looking at the tests in `tensorflow/python/data/experimental/kernel_tests/data_service_ops_test` and didn't find any test which used `distributed_epoch` and raised the `split provider` type error on a ZipDataset. Thus, I thought of adding one. Please correct me if I am wrong.

aaudiber replied:
Ah you're right. The test as it stands is a little misleading though, since it makes it look like the error happens specifically because the datasets inside the zipped dataset are distributed. But in fact any zip-based dataset can't be distributed, so we could simplify the test to use simple datasets (e.g. `dataset_ops.Dataset.range(10)`) inside the zip.

kvignesh1420 replied:
addressed in the commit.

kvignesh1420 committed:
pylint fixes

kvignesh1420 committed:
remove redundant tests

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
+    ds = self.make_distributed_dataset(
+        ds, cluster_2, processing_mode="distributed_epoch")
+
+    error_regex = "Cannot create a split provider for dataset of type DataServiceDataset"  # pylint: disable=line-too-long
```
addressed in the commits.

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
 
+  @combinations.generate(test_base.eager_only_combinations())
+  def testDistributeParallelToDistributed(self):
+
```
addressed in the commits

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -648,6 +648,86 @@ def testRegisterDifferentDatasets(self):
     self.assertNotEqual(id_1.numpy(), id_2.numpy())
 
+  @combinations.generate(test_base.eager_only_combinations())
+  def testDistributeParallelToDistributed(self):
```
addressed in the commits

kvignesh1420 committed:
renamed test for clarity

kvignesh1420 committed:
simplify the test name

kvignesh1420 commented:
@aaudiber I have simplified the tests as per the comments and retained only the ones where we cover the split-provider failures. Updated the description as well.

kvignesh1420 commented on tensorflow/python/data/experimental/kernel_tests/data_service_ops_test.py:
```diff
@@ -680,6 +713,24 @@ def key_func(x):
         self.assertAllEqual(next(it).numpy(), element)
     self.assertEmpty(list(it))
 
+  # @combinations.generate(test_base.eager_only_combinations())
+  # def testCyclicDistribute(self):
+
+  #   cluster_1 = self.create_cluster(num_workers=1)
+  #   cluster_2 = self.create_cluster(num_workers=1)
+  #   num_sizes = 10
+  #   size_repeats = 5
+  #   numbers = [1 * i for i in range(num_sizes)] * size_repeats
+  #   ds = dataset_ops.Dataset.from_tensors(numbers)
+  #   ds = self.make_distributed_dataset(ds, cluster_1, processing_mode="parallel_epochs")
+  #   ds = ds.map(lambda x: x + 1)
+  #   ds = self.make_distributed_dataset(ds, cluster_2, processing_mode="parallel_epochs")
+  #   ds = ds.map(lambda x: x - 1)
+  #   ds = self.make_distributed_dataset(
+  #       ds, cluster_1, processing_mode="parallel_epochs", job_name="temp_jobname")
+
+  #   self.assertDatasetProduces(ds, numbers, assert_items_equal=True)
```
@aaudiber I was trying to experiment on a cyclic distribute using this test and encountered a timeout instead of a failure. Any idea why this is the case? 
P.S: This was just out of curiosity.

aaudiber replied:
This looks like a deadlock triggered when a worker's GetNext request requires calling back to the same worker. Worker's hold a lock during GetNext: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/data/service/worker_impl.cc#L184

I have an unrelated outstanding CL that will relax the locking so that the lock isn't held while calling Iterator::GetNext. That should fix this deadlock

kvignesh1420 replied:
Thanks for the info. Will add that test once the deadlock is handled. Also, I have removed this commented test for final review.

kvignesh1420 committed:
remove cyclic distribute test
